### ISSUE TITLE
Ignoring: Add support for ignoring Map keys

### ISSUE THREAD

**Issue by jatcwang** [softwaremill/diffx#29]
We can use `each` to specify fields to ignore in `Map` values, but not keys.

**Comment by ghostbuster91** [comment]
Hi, you mean ignoring single key from map or ignoring all keys from map which means, treating map like a list of values?

The first one is currently not possible, but if this is your case we can probably add an additional extension e.g. `.key[K](k: K)` for that. What do you think?

The second one can be achieved using `.contramap`:
`def mapWithoutKeys[K,V]:Diff[Map[K,V]]= Derived[Diff[List[V]]].contramap(_.values.toList)`

**Comment by jatcwang** [comment]
I think our use case is a bit more interesting.

This is roughly what we have

```
case class KeyModel(id: UUID, name: String)
case class MyLookup(map: Map[KeyModel, String]
```

Now since `id` field of `KeyModel` is randomly generated, we want to exclude that when comparing two `MyLookup`. Hence I was hoping to be able to "traverse" into the `id` field from MyLookup, something like `Derived[Diff[MyLookup]].ignore(_.map.eachKey.id)`.

Thinking about more, I think we can reuse `Set`'s `Diff` instance. However with the current definition of `Set` derivation it's not very convenient:

```
  implicit def diffForSet[T: ObjectMatcher, C[W] <: scala.collection.Set[W]](
      implicit ddt: Derived[Diff[T]]
  ): Derived[Diff[C[T]]] =
```

We will need to hand-code our `ObjectMapper[KeyModel]` instance. Could it instead take an implicit parameter of type `Diff[T]`? This way we can do this with the above example:

```
implicit val diffKeyModelTuple: Diff[(KeyModel, String)] = Derived[Diff[(KeyModel, String)]].ignore(_._1.id)
implicit val diffForMap: Diff[Map[KeyModel, String]] = Derived[Diff[Set[(KeyModel, String)]]].contramap(_.toSet)
// use Derived[Diff[MyLookup]]
```

**Comment by ghostbuster91** [comment]
That's indeed more interesting :) 
So I looked into map implementation and there were some bugs. For example there was only instance of `Diff` type class for map with string keys. All fixes are in #31 
It should also help in your case. Take a look especially at that test: https://github.com/softwaremill/diffx/pull/31/files#diff-8c9d6cf88962b0773a7c0d15aca4d93bR336
And let me know if this is something what you were looking for :)

**Comment by jatcwang** [comment]
Yep for the identical case it looks good. Do you think we should test the case where it is not identical for `MyLookup`?

**Comment by ghostbuster91** [comment]
Yes, and because keys could be different but still match (because of objectMatcher) I think that we should treat `Map[K,V]` as `Set[(K,V)]` for comparison purpose.

**Comment by jatcwang** [comment]
It's slightly off topic, but is there a need for ObjectMatcher? I feel like `Diff` is strictly more powerful (perhaps Diff should extends ObjectMatcher?)

**Comment by ghostbuster91** [comment]
That's a very good question which I asked myself today :)
I think that there is and here is my explanation:
Given a set of quite complex structures u can define when two instances should be considered as same object despite their differences. This is something which cannot be done using `Diff+ignore.` So in my opinion `ObjectMatcher` and `Diff` complement each other.

**Comment by ghostbuster91** [comment]
0.3.7 released. Let me know if everything works correctly

**Comment by jatcwang** [comment]
Hi @ghostbuster91 it looks better now :) 

```
    implicit val km: ObjectMatcher[KeyModel] = new ObjectMatcher[KeyModel] {
      override def isSameObject(left: KeyModel, right: KeyModel): Boolean = left.name == right.name
    }

    implicit val kmd: Diff[KeyModel] = Derived[Diff[KeyModel]].ignore(_.id)

    val uuid1 = UUID.randomUUID()
    val uuid2 = UUID.randomUUID()

    val a1 = MyLookup(Map(KeyModel(uuid1, "k1") -> "val1"))
    val a2 = MyLookup(Map(KeyModel(uuid2, "k1") -> "val1"))

    a1 should matchTo(a2)
```
Given this example you can see that we're duplicating the logic of "ignore id" in two places and it'll cause even more confusion if we ever add a new field to `KeyModel`.
I think if we merge the functionality of ObjectMatcher into Diff, we can achieve something with less boilerplate.

Here's the idea:

Remove ObjectMatcher and Add `isSameObject` method to Diff, with the default implementation being the equality of all fields.

`Diff#ignore` will change the logic of both `apply` and `isSameLogic` to ignore that field 
(I think this is what ppl want most of the time anyway), and we can later provide methods like `diffIgnore` and `sameObjectIgnore`
for configurability if we need them to be different.

This will reduce the boilerplate down to just `implicit val kmd: Diff[KeyModel] = Derived[Diff[KeyModel]].ignore(_.id)`

YDYT?

**Comment by ghostbuster91** [comment]
Have a look at that test:
```
    it("ignore part of map's key using keys's diff specification") {
      implicit def dm: Diff[KeyModel] = Derived[Diff[KeyModel]].ignore(_.id)

      val a1 = MyLookup(Map(KeyModel(UUID.randomUUID(), "k1") -> "val1"))
      val a2 = MyLookup(Map(KeyModel(UUID.randomUUID(), "k1") -> "val1"))
      compare(a1, a2) shouldBe Identical(a1)
    }
```

You don't have to duplicate logic. 
The purpose of `objectMatcher` is to match instances despite their differences, but their differences will still be taken into account while showing results. Think about comparing sets. You want to be able to pair different objects and still compare them. Having said that, ignoring also affects how objects are matched. That's why it is enough to specify `Diff[KeyModel]` with ignored `id` field to achieve what you showed.

To complete the picture take a look at the following test, where only `objectMatcher` is specified:
```
    it("should match values using object mapper") {
      implicit val om: ObjectMatcher[KeyModel] = new ObjectMatcher[KeyModel] {
        override def isSameObject(left: KeyModel, right: KeyModel): Boolean = left.name == right.name
      }
      val uuid1 = UUID.randomUUID()
      val uuid2 = UUID.randomUUID()
      val a1 = MyLookup(Map(KeyModel(uuid1, "k1") -> "val1"))
      val a2 = MyLookup(Map(KeyModel(uuid2, "k1") -> "val1"))
      compare(a1, a2) shouldBe DiffResultObject(
        "MyLookup",
        Map(
          "map" -> DiffResultMap(
            Map(
              DiffResultObject(
                "KeyModel",
                Map(
                  "id" -> DiffResultValue(uuid1, uuid2),
                  "name" -> Identical("k1")
                )
              ) -> Identical("val1")
            )
          )
        )
      )
    }
  }
```

**Comment by jatcwang** [comment]
I see I'm happy then :) Thanks for the explanation!
